chainstore.go

```diff
@@ -1,147 +1,228 @@
+	defer s.errMu.Unlock()
+	s.errE = err
+}
+
+// Chain represents a store chain.
 type Chain struct {
-	stores []Store
-	async  bool
+	stores []*storeWrapper
 }
 
+// New creates a new store chain backed by the passed stores.
 func New(stores ...Store) Store {
-	return &Chain{stores, false}
-	// TODO: make the chain..
-	// call Open(), but in case of error..?
+	c := &Chain{
+		stores: make([]*storeWrapper, 0, len(stores)),
+	}
+	for _, s := range stores {
+		c.stores = append(c.stores, &storeWrapper{Store: s})
+	}
+	return c
 }
 
-func Async(stores ...Store) Store {
-	return &Chain{stores, true}
-}
+// Open opens all the stores.
+func (c *Chain) Open() error {
 
-func (c *Chain) Open() (err error) {
-	for _, s := range c.stores {
-		err = s.Open()
-		if err != nil {
-			return // return first error that comes up
-		}
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	return
+
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(s *storeWrapper) {
+			defer wg.Done()
+			s.setErr(s.Open())
+		}(c.stores[i])
+	}
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Close() (err error) {
-	for _, s := range c.stores {
-		err = s.Close()
-		// TODO: we shouldn't stop on first error.. should keep trying to close
-		// and record errors separately
-		if err != nil {
-			return
-		}
+// Close closes all the stores.
+func (c *Chain) Close() error {
+	var wg sync.WaitGroup
+
+	for i := range c.stores {
+		wg.Add(1)
+		go func(s *storeWrapper) {
+			defer wg.Done()
+			s.setErr(s.Close())
+		}(c.stores[i])
 	}
-	return
+
+	wg.Wait()
+
+	return c.firstErr()
 }
 
-func (c *Chain) Put(key string, val []byte) (err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Put(ctx context.Context, key string, val []byte) (err error) {
+	if !isValidKey(key) {
 		return ErrInvalidKey
 	}
 
-	fn := func() (err error) {
-		for _, s := range c.stores {
-			err = s.Put(key, val)
-			if err != nil {
-				return
-			}
-		}
-		return
+	if err := c.firstErr(); err != nil {
+		return fmt.Errorf("Open failed due to a previous error: %q", err)
 	}
-	if c.async {
-		go fn()
-	} else {
-		err = fn()
+
+	fn := func(s Store) error {
+		return s.Put(ctx, key, val)
 	}
-	return
+
+	return c.doWithContext(ctx, fn)
 }
 
-func (c *Chain) Get(key string) (val []byte, err error) {
-	if !IsValidKey(key) {
+func (c *Chain) Get(ctx context.Context, key string) (val []byte, err error) {
+	if !isValidKey(key) {
 		return nil, ErrInvalidKey
 	}
 
-	for i, s := range c.stores {
-		val, err = s.Get(key)
-		if err != nil {
-			return
-		}
+	if err := c.firstErr(); err != nil {
+		return nil, fmt.Errorf("Open failed due to a previous error: %q", err)
+	}
 
-		if len(val) > 0 {
-			if i > 0 {
-				// put the value in all other stores up the chain
-				fn := func() {
-					for n := i - 1; n >= 0; n-- {
-						c.stores[n].Put(key, val) // errors..?
-					}
+	nextStore := make(chan Store, len(c.stores))
```

Comment:
Hey @VojtechVitek,

I had to modify your snippet a bit, what do you think of the new location of the range over `putBack`? Otherwise it just stays blocked waiting for a value to arrive.